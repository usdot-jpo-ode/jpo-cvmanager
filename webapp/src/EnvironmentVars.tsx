class EnvironmentVars {
  static getBaseApiUrl() {
    return process.env.REACT_APP_GATEWAY_BASE_URL
  }

  static getMessageTypes() {
    const COUNT_MESSAGE_TYPES = process.env.REACT_APP_COUNT_MESSAGE_TYPES
    if (!COUNT_MESSAGE_TYPES) {
      return []
    }
    const messageTypes = COUNT_MESSAGE_TYPES.split(',').map((item) => item.trim())
    return messageTypes
  }

  static getMessageViewerTypes() {
    const VIEWER_MESSAGE_TYPES = process.env.REACT_APP_VIEWER_MESSAGE_TYPES
    if (!VIEWER_MESSAGE_TYPES) {
      return ['BSM'] // default to BSM if not set
    }
    const messageTypes = VIEWER_MESSAGE_TYPES.split(',').map((item) => item.trim())
    return messageTypes
  }

  static getMapboxInitViewState() {
    const MAPBOX_INIT_LATITUDE = Number(process.env.REACT_APP_MAPBOX_INIT_LATITUDE)
    const MAPBOX_INIT_LONGITUDE = Number(process.env.REACT_APP_MAPBOX_INIT_LONGITUDE)
    const MAPBOX_INIT_ZOOM = Number(process.env.REACT_APP_MAPBOX_INIT_ZOOM)

    const viewState = {
      latitude: MAPBOX_INIT_LATITUDE,
      longitude: MAPBOX_INIT_LONGITUDE,
      zoom: MAPBOX_INIT_ZOOM,
    }

    return viewState
  }

  static MAPBOX_TOKEN = process.env.REACT_APP_MAPBOX_TOKEN
  static CVIZ_API_SERVER_URL = process.env.REACT_APP_CVIZ_API_SERVER_URL
  static CVIZ_API_WS_URL = process.env.REACT_APP_CVIZ_API_WS_URL
  static KEYCLOAK_HOST_URL = process.env.REACT_APP_KEYCLOAK_URL
  static KEYCLOAK_REALM = process.env.REACT_APP_KEYCLOAK_REALM
  static DOT_NAME = process.env.REACT_APP_DOT_NAME

  static rsuInfoEndpoint = `${this.getBaseApiUrl()}/rsuinfo`
  static rsuOnlineEndpoint = `${this.getBaseApiUrl()}/rsu-online-status`
  static rsuCountsEndpoint = `${this.getBaseApiUrl()}/rsucounts`
  static rsuCommandEndpoint = `${this.getBaseApiUrl()}/rsu-command`
  static wzdxEndpoint = `${this.getBaseApiUrl()}/wzdx-feed`
  static rsuMapInfoEndpoint = `${this.getBaseApiUrl()}/rsu-map-info`
  static rsuGeoQueryEndpoint = `${this.getBaseApiUrl()}/rsu-geo-query`
  static rsuMsgFwdQueryEndpoint = `${this.getBaseApiUrl()}/rsu-msgfwd-query`
  static geoMsgDataEndpoint = `${this.getBaseApiUrl()}/rsu-geo-msg-data`
  static issScmsStatusEndpoint = `${this.getBaseApiUrl()}/iss-scms-status`
  static ssmSrmEndpoint = `${this.getBaseApiUrl()}/rsu-ssm-srm-data`
  static authEndpoint = `${this.getBaseApiUrl()}/user-auth`
  static adminAddRsu = `${this.getBaseApiUrl()}/admin-new-rsu`
  static adminRsu = `${this.getBaseApiUrl()}/admin-rsu`
  static adminAddUser = `${this.getBaseApiUrl()}/admin-new-user`
  static adminUser = `${this.getBaseApiUrl()}/admin-user`
  static adminNotification = `${this.getBaseApiUrl()}/admin-notification`
  static adminAddNotification = `${this.getBaseApiUrl()}/admin-new-notification`
  static adminAddOrg = `${this.getBaseApiUrl()}/admin-new-org`
  static adminOrg = `${this.getBaseApiUrl()}/admin-org`
  static contactSupport = `${this.getBaseApiUrl()}/contact-support`
  static rsuErrorSummary = `${this.getBaseApiUrl()}/rsu-error-summary`
}

export default EnvironmentVars
