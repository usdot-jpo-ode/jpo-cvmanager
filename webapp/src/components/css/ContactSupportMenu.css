#ContactSupportMenu {
  position: relative;
  padding: 1rem;
  margin: 1rem;
  border-radius: 1rem;
}

h5 {
  font-size: 1.5rem;
  font-weight: 400;
  margin-bottom: 0.5rem;
  text-align: center;
}

.label {
  font-size: 1rem;
  font-weight: 400;
  margin-bottom: 0.5rem;
}

.btn {
  font-size: 1rem;
  font-weight: 400;
  margin-bottom: 0.5rem;
}

.mb-3 {
  margin-bottom: 1rem !important;
}

textarea {
  resize: vertical;
  min-height: 100px;
}

.contactWrapper {
  position: fixed;
  left: 0;
  bottom: 0;
  display: flex;
  padding: 16px 0px;
  justify-content: center;
  align-items: center;
  width: 100%;
  background-color: rgba(0, 0, 0, 0.5);
}
