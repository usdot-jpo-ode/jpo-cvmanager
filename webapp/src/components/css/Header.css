#header {
  text-align: left;
  z-index: 100;
  width: 100%;
  height: 100px;
}

#frontpagelogo {
  height: 100px;
  width: 250px;
}

#frontgrid {
  margin-top: 20%;
}

#frontpage {
  font-family: Arial, Helvetica, sans-serif;
  font-size: 30px;
  position: absolute;
  min-width: 100%;
  min-height: 100%;
}

#logo {
  padding: 5px;
  align-items: center;
}

#header-text {
  padding: 5px;
  margin-left: 10px;
  align-items: center;
}

#keycloakbtndiv {
  margin-top: 40px;
  font-family: Arial, Helvetica, sans-serif;
  cursor: pointer;
  border-radius: 3px;
}

#login {
  padding: 10px 0px;
  margin-left: 100px;
  position: absolute;
  right: 10px;
}

#loginMessage {
  font-size: 25px;
  padding: 30px;
}

#userInfoGrid {
  padding-right: 30px;
  height: 100px;
}

#nameText {
  font-size: 25px;
  text-align: right;
  margin-top: 5px;
}

#emailText {
  font-size: 18px;
  text-align: right;
}

#roleText {
  font-size: 18px;
  text-align: right;
}

#helpLink {
  margin-left: 1.5em;
  margin-top: 0.5em;
}
