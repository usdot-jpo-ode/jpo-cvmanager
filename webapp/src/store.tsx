import { Action, ThunkAction, configureStore } from '@reduxjs/toolkit'
import rsuReducer from './generalSlices/rsuSlice'
import userReducer from './generalSlices/userSlice'
import wzdxReducer from './generalSlices/wzdxSlice'
import configReducer from './generalSlices/configSlice'
import intersectionReducer from './generalSlices/intersectionSlice'
import adminAddOrganizationReducer from './features/adminAddOrganization/adminAddOrganizationSlice'
import adminAddRsuReducer from './features/adminAddRsu/adminAddRsuSlice'
import adminAddIntersectionReducer from './features/adminAddIntersection/adminAddIntersectionSlice'
import adminAddUserReducer from './features/adminAddUser/adminAddUserSlice'
import adminEditOrganizationReducer from './features/adminEditOrganization/adminEditOrganizationSlice'
import adminEditRsuReducer from './features/adminEditRsu/adminEditRsuSlice'
import adminEditIntersectionReducer from './features/adminEditIntersection/adminEditIntersectionSlice'
import adminEditUserReducer from './features/adminEditUser/adminEditUserSlice'
import adminOrganizationTabReducer from './features/adminOrganizationTab/adminOrganizationTabSlice'
import adminOrganizationTabUserReducer from './features/adminOrganizationTabUser/adminOrganizationTabUserSlice'
import adminOrganizationTabRsuReducer from './features/adminOrganizationTabRsu/adminOrganizationTabRsuSlice'
import adminOrganizationTabIntersectionReducer from './features/adminOrganizationTabIntersection/adminOrganizationTabIntersectionSlice'
import adminRsuTabReducer from './features/adminRsuTab/adminRsuTabSlice'
import adminIntersectionTabReducer from './features/adminIntersectionTab/adminIntersectionTabSlice'
import adminUserTabReducer from './features/adminUserTab/adminUserTabSlice'
import adminNotificationTabReducer from './features/adminNotificationTab/adminNotificationTabSlice'
import adminAddNotificationReducer from './features/adminAddNotification/adminAddNotificationSlice'
import adminEditNotificationReducer from './features/adminEditNotification/adminEditNotificationSlice'
import menuReducer from './features/menu/menuSlice'
import asn1DecoderSlice from './features/intersections/decoder/asn1-decoder-slice'
import intersectionMapReducer from './features/intersections/map/map-slice'
import intersectionMapLayerStyleReducer from './features/intersections/map/map-layer-style-slice'
import dataSelectorReducer from './features/intersections/data-selector/dataSelectorSlice'
import mapSliceReducer from './pages/mapSlice'

export const setupStore = (preloadedState: any) => {
  return configureStore({
    reducer: {
      rsu: rsuReducer,
      user: userReducer,
      wzdx: wzdxReducer,
      config: configReducer,
      intersection: intersectionReducer,
      adminAddOrganization: adminAddOrganizationReducer,
      adminAddRsu: adminAddRsuReducer,
      adminAddIntersection: adminAddIntersectionReducer,
      adminAddUser: adminAddUserReducer,
      adminEditOrganization: adminEditOrganizationReducer,
      adminEditRsu: adminEditRsuReducer,
      adminEditIntersection: adminEditIntersectionReducer,
      adminEditUser: adminEditUserReducer,
      adminOrganizationTab: adminOrganizationTabReducer,
      adminOrganizationTabUser: adminOrganizationTabUserReducer,
      adminOrganizationTabRsu: adminOrganizationTabRsuReducer,
      adminOrganizationTabIntersection: adminOrganizationTabIntersectionReducer,
      adminRsuTab: adminRsuTabReducer,
      adminIntersectionTab: adminIntersectionTabReducer,
      adminUserTab: adminUserTabReducer,
      adminNotificationTab: adminNotificationTabReducer,
      adminAddNotification: adminAddNotificationReducer,
      adminEditNotification: adminEditNotificationReducer,
      menu: menuReducer,
      intersectionMap: intersectionMapReducer,
      intersectionMapLayerStyle: intersectionMapLayerStyleReducer,
      dataSelector: dataSelectorReducer,
      map: mapSliceReducer,
      asn1Decoder: asn1DecoderSlice,
    },
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        thunk: true,
        serializableCheck: false,
        immutableCheck: false,
      }),
    devTools: true,
  })
}

type AppStore = ReturnType<typeof setupStore>
export type AppState = ReturnType<AppStore['getState']>
type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>

export type RootState = ReturnType<ReturnType<typeof setupStore>['getState']>
