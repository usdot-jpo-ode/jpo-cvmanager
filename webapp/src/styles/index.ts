import { alpha, createTheme, Theme } from '@mui/material'
import './fonts/museo-slab.css'
import '../App.css'

declare module '@mui/material/styles' {
  interface PaletteOptions {
    custom: {
      mapLegendBackground: string
      tableHeaderBackground: string
      tableErrorBackground: string
      mapStyleFilePath: string
      mapStyleHasTraffic: boolean
      mapMenuItemBackgroundSelected: string
      mapMenuItemBorderSelected: string
      mapMenuItemHoverSelected: string
      mapMenuItemHoverUnselected: string
      mapMenuBackground: string
      intersectionMapAccordionExpanded: string
      intersectionMapButtonHover: string
      rowActionIcon: string
    }
  }
  interface Palette {
    custom: {
      mapLegendBackground: string
      tableHeaderBackground: string
      tableErrorBackground: string
      mapStyleFilePath: string
      mapStyleHasTraffic: boolean
      mapMenuItemBackgroundSelected: string
      mapMenuItemBorderSelected: string
      mapMenuItemHoverSelected: string
      mapMenuItemHoverUnselected: string
      mapMenuBackground: string
      intersectionMapAccordionExpanded: string
      intersectionMapButtonHover: string
      rowActionIcon: string
    }
  }
}

export const headerTabHeight = 119

export const testTheme = createTheme({
  palette: {
    custom: {
      mapLegendBackground: '#000000',
      tableHeaderBackground: '#000000',
      tableErrorBackground: '#000000',
      mapStyleFilePath: 'mapbox-styles/cdot-dark.json',
      mapStyleHasTraffic: true,
      mapMenuItemBackgroundSelected: '#000000',
      mapMenuItemBorderSelected: '#000000',
      mapMenuItemHoverSelected: '#000000',
      mapMenuItemHoverUnselected: '#000000',
      mapMenuBackground: '#000000',
      intersectionMapAccordionExpanded: '#000000',
      intersectionMapButtonHover: '#000000',
      rowActionIcon: '#000000',
    },
  },
})

// Please note that the light theme is currently not being maintained.
// Light Theme - https://www.realtimecolors.com/?colors=0a1424-e7eef8-213e73-7978d9-4431af&fonts=Inter-Inter
// --text: #0a1424;
// --background: #e7eef8;
// --primary: #213e73;
// --secondary: #7978d9;
// --accent: #4431af;
const themeMainLight = createTheme({
  cssVariables: true,
  typography: {
    fontFamily: '"Trebuchet MS", Arial, Helvetica, sans-serif',
    h1: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h2: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h3: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h4: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h5: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h6: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
  },
  components: {
    MuiTableHead: {
      styleOverrides: {
        root: {
          backgroundColor: '#F3F4F6',
          '.MuiTableCell-root': {
            color: '#374151',
          },
          borderBottom: 'none',
          '& .MuiTableCell-root': {
            borderBottom: 'none',
            fontSize: '12px',
            fontWeight: 600,
            lineHeight: 1,
            letterSpacing: 0.5,
            textTransform: 'uppercase',
          },
          '& .MuiTableCell-paddingCheckbox': {
            paddingTop: 4,
            paddingBottom: 4,
          },
        },
      },
    },
    MuiPaper: {
      styleOverrides: {
        root: {
          backgroundImage: 'unset',
        },
      },
    },
    MuiDivider: {
      styleOverrides: {
        root: {
          borderColor: '#061731',
        },
      },
    },
  },
  palette: {
    mode: 'light',
    primary: {
      main: '#6f69e0',
      light: '#2e3574',
      dark: '#a7a3e7',
      contrastText: '#0b041b',
    },
    secondary: {
      main: '#53aaf1',
      light: '#114875',
      dark: '#92bcde',
      contrastText: '#121212',
    },
    error: {
      main: '#713737',
    },
    success: {
      main: '#315131',
    },
    text: {
      primary: '#121828',
      secondary: '#061731',
      disabled: 'rgba(55, 65, 81, 0.48)',
    },
    divider: '#8f8f8f',
    background: {
      default: '#ececec',
      paper: '#FFFFFF',
    },
    custom: {
      mapLegendBackground: '#c8cfda',
      tableHeaderBackground: '#F3F4F6',
      tableErrorBackground: '#fdc7c7',
      mapStyleFilePath: 'mapbox-styles/main-light.json',
      mapStyleHasTraffic: true,
      mapMenuBackground: '#e0e0e0',
      mapMenuItemBackgroundSelected: '#c4c2c2',
      mapMenuItemBorderSelected: '1px solid black',
      mapMenuItemHoverSelected: '#a19f9f',
      mapMenuItemHoverUnselected: '#ffffff',
      intersectionMapAccordionExpanded: '#d7d7d7',
      intersectionMapButtonHover: '#d7d7d7',
      rowActionIcon: '#9DBDD3',
    },
  },
})

// --text: #dbe5f5;
// --background: #1b1d1f;
// --primary: #4383ad;
// --secondary: #dbe5f5;
// --accent: #614fcd;
const themeMainDark = createTheme({
  cssVariables: true,
  typography: {
    fontFamily: '"Trebuchet MS", Arial, Helvetica, sans-serif',
    h1: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h2: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h3: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h4: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h5: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
    h6: {
      fontFamily: '"museo-slab" Arial Helvetica Sans-Serif',
    },
  },
  components: {
    MuiTableHead: {
      styleOverrides: {
        root: {
          backgroundColor: '#252525',
          borderBottom: 'none',
          '& .MuiTableCell-root': {
            borderBottom: 'none',
            fontSize: '12px',
            fontWeight: 600,
            lineHeight: 1,
            letterSpacing: 0.5,
            textTransform: 'uppercase',
          },
          '& .MuiTableCell-paddingCheckbox': {
            paddingTop: 4,
            paddingBottom: 4,
          },
        },
      },
    },
    MuiPaper: {
      styleOverrides: {
        root: {
          backgroundImage: 'unset',
        },
      },
    },
    MuiDivider: {
      styleOverrides: {
        root: {
          borderColor: '#c7c7c7',
        },
      },
    },
  },
  palette: {
    mode: 'dark',
    primary: {
      main: '#3B7BA5',
      light: '#51a2d6',
      dark: '#3B7BA54D',
      contrastText: '#fff',
    },
    secondary: {
      main: '#dbe5f5',
      light: '#535297',
      dark: '#161563',
      contrastText: '#FAFAFA',
    },
    error: {
      light: '#FF6D57',
      main: '#E94F3766',
      dark: '#6e312a',
    },
    success: {
      light: '#75BD27',
      main: '#A0D36466',
      dark: '#9e0e0e',
    },
    text: {
      primary: '#FFFFFF',
      secondary: '#c7c7c7',
      disabled: '#acacac',
    },
    info: {
      main: '#dbe5f5',
    },
    custom: {
      mapLegendBackground: '#1b1d1f',
      tableHeaderBackground: '#252525',
      tableErrorBackground: '#4d2e2e',
      mapStyleFilePath: 'mapbox-styles/main-dark.json',
      mapStyleHasTraffic: true,
      mapMenuBackground: '#3c3c3c',
      mapMenuItemBackgroundSelected: '#333333',
      mapMenuItemBorderSelected: '1px solid black',
      mapMenuItemHoverSelected: '#333333',
      mapMenuItemHoverUnselected: '#575757',
      intersectionMapAccordionExpanded: '#2E2F31',
      intersectionMapButtonHover: '#2D5F7F',
      rowActionIcon: '#9DBDD3',
    },
    background: {
      paper: '#1b1d1f',
      default: '#333333',
    },
  },
})

// All available themes
export const THEMES = {
  light: themeMainLight,
  dark: themeMainDark,
}

export const getCurrentTheme = (isDarkTheme: boolean, lightThemeName: string, darkThemeName: string): Theme => {
  // Warnings to user if theme names are not known
  if (darkThemeName && !THEMES[darkThemeName]) {
    console.warn(`Unknown dark theme name: ${darkThemeName}`)
  }
  if (lightThemeName && !THEMES[lightThemeName]) {
    console.warn(`Unknown light theme name: ${lightThemeName}`)
  }
  if (isDarkTheme) {
    return THEMES[darkThemeName] ?? THEMES.dark
  } else {
    return THEMES[lightThemeName] ?? THEMES.light
  }
}
