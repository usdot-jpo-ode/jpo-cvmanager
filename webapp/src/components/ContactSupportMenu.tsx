import React, { useState } from 'react'
import { Form } from 'react-bootstrap'
import { useForm } from 'react-hook-form'

import 'react-widgets/styles.css'
import RsuApi from '../apis/rsu-api'

import './css/ContactSupportMenu.css'
import toast from 'react-hot-toast'
import Dialog from '@mui/material/Dialog'
import { Button, DialogActions, DialogContent, DialogTitle } from '@mui/material'
import { AdminButton } from '../styles/components/AdminButton'
import { LightButton } from '../styles/components/LightButton'
import ContactSupportIcon from '@mui/icons-material/ContactSupport'

const ContactSupportMenu = () => {
  const [hidden, setHidden] = useState(true) // hidden by default
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm()

  const onSubmit = async (data: Object) => {
    try {
      const res = await RsuApi.postContactSupport(data)
      const status = res.status
      if (status === 200) {
        toast.success('Successfully sent email')
        reset()
      } else {
        toast.error('Something went wrong: ' + status)
      }
    } catch (exception_var) {
      toast.error('An exception occurred, please try again later')
    }
    setHidden(true)
  }

  if (hidden) {
    return (
      <div>
        <LightButton
          variant="contained"
          onClick={() => {
            setHidden(!hidden)
          }}
          sx={{ margin: 1, top: 5 }}
        >
          Contact Support
        </LightButton>
      </div>
    )
  }

  return (
    <Dialog open={true}>
      <DialogTitle>Contact Support</DialogTitle>
      <DialogContent>
        <Form
          id="contact-support-form"
          onSubmit={handleSubmit(onSubmit)}
          style={{ fontFamily: 'Arial, Helvetica, sans-serif' }}
        >
          <Form.Group className="mb-3" controlId="email">
            <Form.Label className="label">Your Email</Form.Label>
            <Form.Control
              type="email"
              placeholder="Enter your email (Required)"
              {...register('email', {
                required: 'Email is required',
              })}
            />
            {errors.email && <Form.Text className="text-danger">{errors.email.message}</Form.Text>}
          </Form.Group>
          <Form.Group className="mb-3" controlId="subject">
            <Form.Label className="label">Subject</Form.Label>
            <Form.Control
              type="text"
              placeholder="Enter your subject (Required)"
              {...register('subject', {
                required: 'Subject is required',
              })}
            />
            {errors.subject && <Form.Text className="text-danger">{errors.subject.message}</Form.Text>}
          </Form.Group>
          <Form.Group className="mb-3" controlId="message">
            <Form.Label className="label">Message</Form.Label>
            <Form.Control
              as="textarea"
              rows={5}
              placeholder="Enter your message (Required)"
              {...register('message', {
                required: 'Message is required',
              })}
            />
            {errors.message && <Form.Text className="text-danger">{errors.message.message}</Form.Text>}
          </Form.Group>
        </Form>
      </DialogContent>
      <DialogActions>
        <AdminButton
          onClick={() => {
            setHidden(!hidden)
          }}
        >
          Close
        </AdminButton>
        <AdminButton form="contact-support-form" type="submit">
          Send Email
        </AdminButton>
      </DialogActions>
    </Dialog>
  )
}

export default ContactSupportMenu
