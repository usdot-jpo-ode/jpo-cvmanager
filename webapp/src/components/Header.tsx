import React, { useEffect, useMemo } from 'react'
import Grid2 from '@mui/material/Grid2'
import EnvironmentVars from '../EnvironmentVars'
import {
  selectOrganizationName,
  selectName,
  selectEmail,
  selectAuthLoginData,
  selectLoginFailure,
  selectKcFailure,

  // actions
  logout,
  changeOrganization,
  setKcFailure,
  selectLoginMessage,
} from '../generalSlices/userSlice'
import { useKeycloak } from '@react-keycloak/web'

import './css/Header.css'

import ContactSupportMenu from './ContactSupportMenu'
import { useAppDispatch, useAppSelector } from '../hooks'
import { AnyAction, ThunkDispatch } from '@reduxjs/toolkit'
import { RootState } from '../store'
import { Button, FormControl, InputLabel, MenuItem, Paper, Select, useTheme } from '@mui/material'
import LogoutIcon from '@mui/icons-material/Logout'
import { LightButton } from '../styles/components/LightButton'

const Header = () => {
    const dispatch = useAppDispatch()
  const theme = useTheme()
  const { keycloak } = useKeycloak()

  const authLoginData = useAppSelector(selectAuthLoginData)
  const organizationName = useAppSelector(selectOrganizationName)
  const userName = useAppSelector(selectName)
  const userEmail = useAppSelector(selectEmail)
  const loginFailure = useAppSelector(selectLoginFailure)
  const kcFailure = useAppSelector(selectKcFailure)
  const loginMessage = useAppSelector(selectLoginMessage)

  const iconPath = useMemo(() => {
    return theme.palette.mode === 'dark' ? '/icons/logo_dark.png' : '/icons/logo_light.png'
  }, [theme.palette.mode])

  useEffect(() => {
    const kcFailureDelay = 500000
    const kcFailureTimer = setTimeout(() => {
      if (!keycloak?.authenticated) {
        console.debug('Login failure logic: User is not authenticated with Keycloak')
        dispatch(setKcFailure(true))
      } else {
        console.debug('Login failure logic: User is now authenticated with Keycloak')
        dispatch(setKcFailure(false))
      }
    }, kcFailureDelay)

    return () => clearTimeout(kcFailureTimer)
  }, [keycloak, keycloak?.authenticated, dispatch])

  const handleUserLogout = () => {
    console.debug('handleUserLogout')
    dispatch(logout())
    keycloak?.logout()
  }

  return (
    <div>
      {authLoginData && keycloak?.authenticated ? (
        <Paper id="header">
          <Grid2 container alignItems="center" style={{ height: 'fit-content' }}>
            <img id="logo" src={iconPath} alt="Logo" height="90px" />
            <h1 id="header-text">{EnvironmentVars.DOT_NAME} CV Manager</h1>
            <div id="login" style={{ lineHeight: 1.1, marginTop: 10 }}>
              <Grid2 container alignItems="center">
                <Grid2 id="userInfoGrid">
                  <h3 id="nameText">{userName}</h3>
                  <h3 id="emailText">{userEmail}</h3>
                  <FormControl sx={{ mt: 0.2, minWidth: 200 }} size="small">
                    <Select
                      value={organizationName}
                      onChange={(event) => dispatch(changeOrganization(event.target.value))}
                    >
                      {(authLoginData?.data?.organizations ?? []).map((permission) => (
                        <MenuItem key={permission.name + 'Option'} value={permission.name} color="primary">
                          {permission.name} ({permission.role})
                        </MenuItem>
                      ))}
                    </Select>
                  </FormControl>
                </Grid2>
                <LightButton
                  startIcon={<LogoutIcon />}
                  onClick={() => handleUserLogout()}
                  sx={{
                    padding: 1,
                    paddingLeft: 2,
                    paddingRight: 2,
                    right: '10px',
                  }}
                >
                  Logout
                </LightButton>
              </Grid2>
            </div>
          </Grid2>
        </Paper>
      ) : (
        <Paper id="frontpage">
          <Grid2 container id="frontgrid" alignItems="center" direction="column">
            <Grid2 container justifyContent="center" alignItems="center">
              <img id="frontpagelogo" src={iconPath} alt="Logo" height="90px" />
              <h1 id="header-text">{EnvironmentVars.DOT_NAME} CV Manager</h1>
            </Grid2>
            {loginFailure && <h3 id="loginMessage">{loginMessage}</h3>}
            <div id="keycloakbtndiv">
              {loginFailure && (
                <Button variant="contained" onClick={() => handleUserLogout()}>
                  Logout User
                </Button>
              )}
            </div>
            {kcFailure && <h3 id="loginMessage">Application Authentication Error!</h3>}

            <br />

            {loginFailure && <ContactSupportMenu />}
          </Grid2>
        </Paper>
      )}
    </div>
  )
}

export default Header