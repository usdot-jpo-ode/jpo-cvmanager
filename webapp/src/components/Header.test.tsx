import React from 'react'
import { render } from '@testing-library/react'
import Header from './Header'
import { Provider } from 'react-redux'
import { ThemeProvider } from '@mui/material'
import { testTheme } from '../styles'
import { setupStore } from '../store'
import EnvironmentVars from '../EnvironmentVars'
import { useKeycloak } from '@react-keycloak/web'
import { replaceChaoticIds } from '../utils/test-utils'
import ContactSupportMenu from './ContactSupportMenu'

jest.mock('@react-keycloak/web')

const mockKeycloak = {
  authenticated: true,
  login: jest.fn(),
  logout: jest.fn(),
  register: jest.fn(),
  accountManagement: jest.fn(),
  loadUserProfile: jest.fn(),
}

describe('<Header />', () => {
  beforeEach(() => {
    ;(useKeycloak as any).mockReturnValue([mockKeycloak])
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should take a snapshot', () => {
    const { container } = render(
      <ThemeProvider theme={testTheme}>
        <Provider
          store={setupStore({
            user: {
              value: {
                loginFailure: true,
                loginMessage: 'User Unauthorized',
                atuhLoginData: {
                  data: {
                    organizations: [{ name: 'org1', role: 'role1' }],
                  },
                },
              },
            },
          })}
        >
          <Header />
          <br />
          <ContactSupportMenu />
        </Provider>
      </ThemeProvider>
    )

    expect(replaceChaoticIds(container)).toMatchSnapshot()
  })
})
