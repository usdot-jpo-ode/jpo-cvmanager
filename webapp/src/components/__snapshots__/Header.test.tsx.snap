// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should take a snapshot 1`] = `
<div>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-mocked-MuiPaper-root"
      id="frontpage"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-column css-mocked-MuiGrid2-root"
        id="frontgrid"
      >
        <div
          class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row css-mocked-MuiGrid2-root"
        >
          <img
            alt="Logo"
            id="frontpagelogo"
            src="/icons/logo_light.png"
          />
          <h1
            id="header-text"
          >
             CV Manager
          </h1>
        </div>
        <h3
          id="loginMessage"
        >
          User Unauthorized
        </h3>
        <div
          id="keycloakbtndiv"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-mocked-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Logout User
          </button>
        </div>
        <br />
        <div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-mocked-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-mocked-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mocked-MuiSvgIcon-root"
                data-testid="ContactSupportIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11.5 2C6.81 2 3 5.81 3 10.5S6.81 19 11.5 19h.5v3c4.86-2.34 8-7 8-11.5C20 5.81 16.19 2 11.5 2m1 14.5h-2v-2h2zm0-3.5h-2c0-3.25 3-3 3-5 0-1.1-.9-2-2-2s-2 .9-2 2h-2c0-2.21 1.79-4 4-4s4 1.79 4 4c0 2.5-3 2.75-3 5"
                />
              </svg>
            </span>
            Contact Support
          </button>
        </div>
      </div>
    </div>
  </div>
  <br />
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-mocked-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-mocked-MuiButton-startIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mocked-MuiSvgIcon-root"
          data-testid="ContactSupportIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.5 2C6.81 2 3 5.81 3 10.5S6.81 19 11.5 19h.5v3c4.86-2.34 8-7 8-11.5C20 5.81 16.19 2 11.5 2m1 14.5h-2v-2h2zm0-3.5h-2c0-3.25 3-3 3-5 0-1.1-.9-2-2-2s-2 .9-2 2h-2c0-2.21 1.79-4 4-4s4 1.79 4 4c0 2.5-3 2.75-3 5"
          />
        </svg>
      </span>
      Contact Support
    </button>
  </div>
</div>
`;
